sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala

```diff
@@ -68,5 +69,13 @@ private case object OracleDialect extends JdbcDialect {
     case _ => None
   }
 
+  override def beforeFetch(connection: Connection, properties: Map[String, String]): Unit = {
+    // Set general date and timestamp format before query.
+    val stmt = connection.createStatement()
+    stmt.execute("alter session set NLS_DATE_FORMAT = 'YYYY-MM-DD'")
+    stmt.execute("alter session set NLS_TIMESTAMP_FORMAT = 'YYYY-MM-DD HH24:MI:SS.FF'")
```

Comment:
To avoid this issue, we should change the way we generate the where clause. That is why I proposed to make `compileValue` Dialect specific. It sounds like there are two approaches to convert the timestamp string to Oracle-recognized timestamp string in the where clause 

- `INSERT INTO datetime VALUES (1, {d '1991-11-09'}, {ts '1996-01-01 01:23:45'}`
- call `to_timestamp_tz`